art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -0,0 +1,389 @@
+        best_indices_poison = None
+
+        if len(np.shape(y_train)) == 2:
+            y_train_classes = np.argmax(y_train, axis=-1)
+        else:
+            y_train_classes = y_train
+        for _ in trange(self.max_trials):
+            if self.selection_strategy == "random":
+                self.indices_poison = np.random.permutation(
+                    np.where([y in classes_target for y in y_train_classes])[0]
+                )[:num_poison_samples]
+            else:
+                self.indices_poison = self.select_poison_indices(
+                    self.substitute_classifier, x_train, y_train, num_poison_samples
+                )
+            x_poison = x_train[self.indices_poison]
+            y_poison = y_train[self.indices_poison]
+            initializer(x_trigger, y_trigger, x_poison, y_poison)
+            original_epochs = self.max_epochs
+            if self.model_retrain:
+                retrain_epochs = self.max_epochs // self.retraining_factor
+                for i in range(self.retraining_factor):
+                    if i == self.retraining_factor - 1:
+                        self.max_epochs = original_epochs - retrain_epochs * i
+                        x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+                    else:
+                        self.max_epochs = retrain_epochs
+                        x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+                        self.model_retraining(x_poisoned)
+            else:
+                x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+            finish_poisoning()
+            B_ = np.mean(B_)  # Averaging B losses from multiple batches.  # pylint: disable=C0103
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = self.indices_poison
+
+        if self.verbose > 0:
+            print("Best B-score:", best_B)
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train
+
+    def get_poison_indices(self):
+        """
+        :return: indices of best poison index
+        """
+        return self.indices_poison
+
+    def model_retraining(self, poisoned_samples):
+        """
+        Applies retraining to substitute model
+        :param poisoned_samples: poisoned array.
+        """
+        from art.utils import load_cifar10
+        from art.estimators.classification.pytorch import PyTorchClassifier
+
+        (x_train, y_train), (x_test, y_test), min_, max_ = load_cifar10()
+        mean = np.mean(x_train, axis=(0, 1, 2, 3))
+        std = np.std(x_train, axis=(0, 1, 2, 3))
+        x_train = (x_train - mean) / (std + 1e-7)
+        x_test = (x_test - mean) / (std + 1e-7)
+        min_ = (min_ - mean) / (std + 1e-7)
+        max_ = (max_ - mean) / (std + 1e-7)
+        x_train = np.transpose(x_train, [0, 3, 1, 2])
```

Comment:
This method is specific for CIFAR10. How about if we ask for the training data as input?

Reply:
So, you mean we pass some variable - dataset_name='cifar10' in this function? I was trying to pass the data from the notebook - but was getting memory error - so is doing it here

Reply:
Yes, passing the data similar as for method `poison`.